File: lib/src/main/java/imajine/Adjustments.java
package imajine;

public class Adjustments {
    /**
     * Adjusts the brightness of the given image by applying a brightness factor to
     * each pixel.
     * The brightness factor should be in the range [-1.0, 1.0], where:
     * <ul>
     * <li>-1.0 represents the darkest adjustment (completely black).</li>
     * <li>0.0 represents no adjustment (original brightness).</li>
     * <li>1.0 represents the brightest adjustment (maximum brightness).</li>
     * </ul>
     *
     * @param image            The image to adjust. This must be an instance of
     *                         {@code Imajine}.
     * @param brightnessFactor The factor by which to adjust the brightness. Values
     *                         outside the range [-1.0, 1.0] will be clamped to this
     *                         range.
     * @throws IllegalArgumentException if the image is null.
     */
    public static void adjustBrightness(Imajine image, double brightnessFactor) {
        // Ensure brightnessFactor is within the valid range
        brightnessFactor = Math.max(-1.0, Math.min(1.0, brightnessFactor));

        // Iterate over each pixel in the image
        for (int y = 0; y < image.getHeight(); y++) {
            for (int x = 0; x < image.getWidth(); x++) {
                // Get the current pixel
                Pixel pixel = image.getPixel(x, y);

                // Adjust the RGB values based on the brightness factor
                int r = adjustColorComponent(pixel.getRed(), brightnessFactor);
                int g = adjustColorComponent(pixel.getGreen(), brightnessFactor);
                int b = adjustColorComponent(pixel.getBlue(), brightnessFactor);

                // Create a new pixel with the adjusted RGB values
                Pixel adjustedPixel = new Pixel(x, y, r, g, b);

                // Set the adjusted pixel back to the image
                image.setPixel(adjustedPixel);
            }
        }
    }

    /**
     * Adjusts the contrast of the given image by applying a contrast factor.
     * The contrast factor determines the intensity of the contrast adjustment,
     * where values range from -1.0 (minimum contrast) to 1.0 (maximum contrast).
     *
     * <p>
     * The method iterates over each pixel in the image, adjusts its RGB
     * components based on the contrast factor, and updates the image with the
     * modified pixel values.
     * </p>
     *
     * @param image          The image to adjust. Must be an instance of
     *                       {@code Imajine}.
     * @param contrastFactor The contrast adjustment factor. Values should be
     *                       between -1.0 and 1.0. Values outside this range
     *                       will be clamped to the nearest valid value.
     * @throws IllegalArgumentException if the image is null.
     */
    public static void adjustContrast(Imajine image, double contrastFactor) {
        // Ensure contrastFactor is within a reasonable range (e.g., -1.0 to 1.0)
        contrastFactor = Math.max(-1.0, Math.min(1.0, contrastFactor));

        // Precompute the scaling factor for contrast adjustment
        double scalingFactor = (1.0 + contrastFactor) / (1.0 - contrastFactor);

        // Iterate over each pixel in the image
        for (int y = 0; y < image.getHeight(); y++) {
            for (int x = 0; x < image.getWidth(); x++) {
                // Get the current pixel
                Pixel pixel = image.getPixel(x, y);

                // Adjust the RGB values for contrast
                int r = adjustContrastComponent(pixel.getRed(), scalingFactor);
                int g = adjustContrastComponent(pixel.getGreen(), scalingFactor);
                int b = adjustContrastComponent(pixel.getBlue(), scalingFactor);

                // Create a new pixel with the adjusted RGB values
                Pixel adjustedPixel = new Pixel(x, y, r, g, b);

                // Set the adjusted pixel back to the image
                image.setPixel(adjustedPixel);
            }
        }
    }

    /**
     * Applies a threshold filter to the given image. Each pixel in the image is
     * converted to grayscale and compared against the specified threshold value. If
     * the grayscale value of a pixel is less than the threshold, the pixel is set
     * to black (0, 0, 0). Otherwise, the pixel is set to white (255, 255, 255).
     *
     * @param image     The image to which the threshold filter will be applied. The
     *                  image should support methods for retrieving and setting
     *                  pixel values.
     * @param threshold The threshold value used to determine whether a pixel is set
     *                  to black or white. This value should be in the range [0,
     *                  255].
     */
    public static void threshold(Imajine image, int threshold) {
        // Iterate over each pixel in the image
        for (int y = 0; y < image.getHeight(); y++) {
            for (int x = 0; x < image.getWidth(); x++) {
                // Get the current pixel
                Pixel pixel = image.getPixel(x, y);

                // Calculate the grayscale value of the pixel
                int gray = (int) (0.299 * pixel.getRed() + 0.587 * pixel.getGreen() + 0.114 * pixel.getBlue());

                // Set the pixel to black or white based on the threshold
                int value = gray < threshold ? 0 : 255;

                // Create a new pixel with the thresholded value
                Pixel thresholdedPixel = new Pixel(x, y, value, value, value);

                // Set the thresholded pixel back to the image
                image.setPixel(thresholdedPixel);
            }
        }
    }

    /**
     * Inverts the colors of the given image. Each pixel's RGB values are inverted
     * by subtracting them from 255.
     *
     * @param image The image to be inverted. This must be an instance of
     *              {@code Imajine}.
     */
    public static void invert(Imajine image) {
        // Iterate over each pixel in the image
        for (int y = 0; y < image.getHeight(); y++) {
            for (int x = 0; x < image.getWidth(); x++) {
                // Get the current pixel
                Pixel pixel = image.getPixel(x, y);

                // Invert the RGB values
                int r = 255 - pixel.getRed();
                int g = 255 - pixel.getGreen();
                int b = 255 - pixel.getBlue();

                // Create a new pixel with the inverted RGB values
                Pixel invertedPixel = new Pixel(x, y, r, g, b);

                // Set the inverted pixel back to the image
                image.setPixel(invertedPixel);
            }
        }
    }

    /**
     * Desaturates the colors of the given image by converting each pixel to its
     * grayscale equivalent. The grayscale value is calculated using the luminosity
     * method.
     *
     * @param image The image to be desaturated. This must be an instance of
     *              {@code Imajine}.
     */
    public static void desaturate(Imajine image) {
        // Iterate over each pixel in the image
        for (int y = 0; y < image.getHeight(); y++) {
            for (int x = 0; x < image.getWidth(); x++) {
                // Get the current pixel
                Pixel pixel = image.getPixel(x, y);

                // Calculate the grayscale value of the pixel
                int gray = (int) (0.299 * pixel.getRed() + 0.587 * pixel.getGreen() + 0.114 * pixel.getBlue());

                // Create a new pixel with the grayscale value
                Pixel desaturatedPixel = new Pixel(x, y, gray, gray, gray);

                // Set the desaturated pixel back to the image
                image.setPixel(desaturatedPixel);
            }
        }
    }

    /**
     * Posterizes the colors of the given image by reducing the number of color
     * levels in each RGB channel. The number of levels is specified by the
     * {@code levels} parameter.
     *
     * @param image  The image to be posterized. This must be an instance of
     *               {@code Imajine}.
     * @param levels The number of color levels to use for posterization. This
     *               should be a positive integer greater than 1.
     */
    public static void posterize(Imajine image, int levels) {
        // Iterate over each pixel in the image
        for (int y = 0; y < image.getHeight(); y++) {
            for (int x = 0; x < image.getWidth(); x++) {
                // Get the current pixel
                Pixel pixel = image.getPixel(x, y);

                // Posterize the RGB values
                int r = posterizeComponent(pixel.getRed(), levels);
                int g = posterizeComponent(pixel.getGreen(), levels);
                int b = posterizeComponent(pixel.getBlue(), levels);

                // Create a new pixel with the posterized RGB values
                Pixel posterizedPixel = new Pixel(x, y, r, g, b);

                // Set the posterized pixel back to the image
                image.setPixel(posterizedPixel);
            }
        }
    }

    /**
     * Posterizes a single color component by reducing its number of levels.
     *
     * @param colorComponent The color component to posterize (0-255).
     * @param levels         The number of levels to use for posterization.
     * @return The posterized color component.
     */
    private static int posterizeComponent(int colorComponent, int levels) {
        // Divide the color component into 4 equal intervals
        int interval = 256 / levels;

        // Determine the posterized value based on the interval
        int value = (colorComponent / interval) * interval;

        // Ensure the value is within the valid range (0-255)
        return Math.max(0, Math.min(255, value));
    }

    /**
     * Adjusts a single color component based on the brightness factor.
     *
     * @param colorComponent   The color component to adjust (0-255).
     * @param brightnessFactor The brightness adjustment factor (-1.0 to 1.0).
     * @return The adjusted color component.
     */
    private static int adjustColorComponent(int colorComponent, double brightnessFactor) {
        // Adjust the color component based on the brightness factor
        int adjustedComponent = (int) (colorComponent * (1 + brightnessFactor));

        // Ensure the adjusted component is within the valid range (0-255)
        return Math.max(0, Math.min(255, adjustedComponent));
    }

    /**
     * Adjusts a single color component based on the contrast scaling factor.
     *
     * @param colorComponent The color component to adjust (0-255).
     * @param scalingFactor  The contrast scaling factor.
     * @return The adjusted color component.
     */
    private static int adjustContrastComponent(int colorComponent, double scalingFactor) {
        // Normalize the color component to the range [-128, 128]
        double normalizedComponent = colorComponent - 128;

        // Apply the contrast scaling factor
        double adjustedComponent = normalizedComponent * scalingFactor;

        // Denormalize the color component back to the range [0, 255]
        int result = (int) (adjustedComponent + 128);

        // Clamp the result to ensure it stays within the valid range
        return Math.max(0, Math.min(255, result));
    }
}

File: lib/src/main/kotlin/imajine/Adjustments.kt
package imajine

/**
 * The Adjustments class provides various image adjustment operations such as brightness, contrast, thresholding, inversion, desaturation, posterization, and photo filters.
 */
object Adjustments {

    /**
     * Adjusts the brightness of the given image by applying a brightness factor to each pixel.
     *
     * @param image The image to adjust.
     * @param brightnessFactor The factor by which to adjust the brightness. Clamped to [-1.0, 1.0].
     */
    fun adjustBrightness(image: Imajine, brightnessFactor: Double) {
        val clampedFactor = brightnessFactor.coerceIn(-1.0, 1.0)
        for (y in 0 until image.height) {
            for (x in 0 until image.width) {
                val pixel = image.getPixel(x, y)
                val r = adjustColorComponent(pixel.red, clampedFactor)
                val g = adjustColorComponent(pixel.green, clampedFactor)
                val b = adjustColorComponent(pixel.blue, clampedFactor)
                image.setPixel(Pixel(x, y, r, g, b))
            }
        }
    }

    /**
     * Adjusts the contrast of the given image by applying a contrast factor.
     *
     * @param image The image to adjust.
     * @param contrastFactor The contrast adjustment factor. Clamped to [-1.0, 1.0].
     */
    fun adjustContrast(image: Imajine, contrastFactor: Double) {
        val clampedFactor = contrastFactor.coerceIn(-1.0, 1.0)
        val scalingFactor = (1.0 + clampedFactor) / (1.0 - clampedFactor)
        for (y in 0 until image.height) {
            for (x in 0 until image.width) {
                val pixel = image.getPixel(x, y)
                val r = adjustContrastComponent(pixel.red, scalingFactor)
                val g = adjustContrastComponent(pixel.green, scalingFactor)
                val b = adjustContrastComponent(pixel.blue, scalingFactor)
                image.setPixel(Pixel(x, y, r, g, b))
            }
        }
    }

    /**
     * Applies a threshold filter to the given image.
     *
     * @param image The image to adjust.
     * @param threshold The threshold value (0-255).
     */
    fun threshold(image: Imajine, threshold: Int) {
        for (y in 0 until image.height) {
            for (x in 0 until image.width) {
                val pixel = image.getPixel(x, y)
                val gray = (0.299 * pixel.red + 0.587 * pixel.green + 0.114 * pixel.blue).toInt()
                val value = if (gray < threshold) 0 else 255
                image.setPixel(Pixel(x, y, value, value, value))
            }
        }
    }

    /**
     * Inverts the colors of the given image.
     *
     * @param image The image to adjust.
     */
    fun invert(image: Imajine) {
        for (y in 0 until image.height) {
            for (x in 0 until image.width) {
                val pixel = image.getPixel(x, y)
                val r = 255 - pixel.red
                val g = 255 - pixel.green
                val b = 255 - pixel.blue
                image.setPixel(Pixel(x, y, r, g, b))
            }
        }
    }

    /**
     * Desaturates the colors of the given image by converting each pixel to grayscale.
     *
     * @param image The image to adjust.
     */
    fun desaturate(image: Imajine) {
        for (y in 0 until image.height) {
            for (x in 0 until image.width) {
                val pixel = image.getPixel(x, y)
                val gray = (0.299 * pixel.red + 0.587 * pixel.green + 0.114 * pixel.blue).toInt()
                image.setPixel(Pixel(x, y, gray, gray, gray))
            }
        }
    }

    /**
     * Posterizes the colors of the given image by reducing the number of color levels.
     *
     * @param image The image to adjust.
     * @param levels The number of color levels (greater than 1).
     */
    fun posterize(image: Imajine, levels: Int) {
        for (y in 0 until image.height) {
            for (x in 0 until image.width) {
                val pixel = image.getPixel(x, y)
                val r = posterizeComponent(pixel.red, levels)
                val g = posterizeComponent(pixel.green, levels)
                val b = posterizeComponent(pixel.blue, levels)
                image.setPixel(Pixel(x, y, r, g, b))
            }
        }
    }

    /**
     * Applies a photo filter to the given image by blending each pixel with a specified color.
     *
     * @param image The image to adjust.
     * @param r The red component of the filter color (0-255).
     * @param g The green component of the filter color (0-255).
     * @param b The blue component of the filter color (0-255).
     * @param density The blending density (0.0 to 1.0).
     */
    fun applyPhotoFilter(image: Imajine, r: Int, g: Int, b: Int, density: Float) {
        val clampedDensity = density.coerceIn(0.0f, 1.0f)
        for (y in 0 until image.height) {
            for (x in 0 until image.width) {
                val pixel = image.getPixel(x, y)
                val alteredR = blendColor(pixel.red, r, clampedDensity)
                val alteredG = blendColor(pixel.green, g, clampedDensity)
                val alteredB = blendColor(pixel.blue, b, clampedDensity)
                image.setPixel(Pixel(x, y, alteredR, alteredG, alteredB))
            }
        }
    }

    private fun adjustColorComponent(colorComponent: Int, brightnessFactor: Double): Int {
        val adjusted = (colorComponent * (1 + brightnessFactor)).toInt()
        return adjusted.coerceIn(0, 255)
    }

    private fun adjustContrastComponent(colorComponent: Int, scalingFactor: Double): Int {
        val normalized = colorComponent - 128
        val adjusted = (normalized * scalingFactor).toInt() + 128
        return adjusted.coerceIn(0, 255)
    }

    private fun posterizeComponent(colorComponent: Int, levels: Int): Int {
        val interval = 256 / levels
        return (colorComponent / interval) * interval
    }

    private fun blendColor(original: Int, filter: Int, density: Float): Int {
        return (original * (1 - density) + filter * density).toInt()
    }
}
